Windows Installation/Deployment Guide

shieldsjared opened the pull request:
GitHub Issue #360 

A rough draft ready for review and merge.  A couple of things worth noting:
- [PLT-2068](https://mattermost.atlassian.net/browse/PLT-4068): Windows does not come with a utility to decompress a .tar.gz file. 
- [PLT-4110](https://mattermost.atlassian.net/browse/PLT-4110): Opening config.json with notepad (default) results in a malformed display due to the type of linefeeds used. 
- Updated the MySQL version requirement to 5.7 because that is the latest and the instructions are based on this version.

The two jira issues will make the process easier, and once completed, the guide can be updated accordingly; however, the guide currently accounts/notates these issues.

Let me know what you think!

esethna commented:
This is awesome, thanks!! Starting review process

it33 commented:
Opening ticket for dev review of this, since it probably needs to be scheduled: https://mattermost.atlassian.net/browse/PLT-4235

yangchen1 commented:
+1. This is awesome.

it33 commented:
@esethna is there a ticket tracking the config.json changes needed for the file to display properly in Notepad? https://mattermost.atlassian.net/browse/PLT-4235

Maybe combine it with the .zip ticket?

esethna commented:
Yep, I've combined the tickets here: https://mattermost.atlassian.net/browse/PLT-4068

shieldsjared commented:
@it33 @yangchen1 @esethna - Should I go ahead and remove the IIS section in anticipation of 3.5? (only needed for SSL).  Are the TLS changes feature-complete in master?  If so, I'll test it out on windows.

it33 commented:
For production deployment IIS would still be needed since Mattermost doesn't yet have web server style logging. Woulf you be interested in the ticket for making configuration.json work in Notepad? It's one the last two tickets we need to get this out...

it33 commented:
Last message is weird as I am on a phone with weird auto-complete

shieldsjared commented:
@it33 - Just [submitted a PR](https://github.com/mattermost/platform/pull/4219) to handle the ZIP and the CRLF for the `config.json`.

shieldsjared commented:
@it33 - Aside from fixing the small conflict, what else can be done to push this along?  The impediments to the deployment process for windows have been resolved.

it33 commented:
Thanks Jared! Sorry for the delay, with the fixes in we just need someone to test it. Let me post in developers and if we can't find someone there we might prioritize a ticket

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+.. attention :: This install guide has been contributed by the Mattermost community. 
+   It has not yet been tested by the core team. We have `an open ticket 
+   <https://github.com/mattermost/docs/issues/360>`__ requesting community help testing and 
+   improving this guide. Once the community has confirmed we have multiple deployments on these 
```
`requesting community help testing and improving this guide` --> `requesting community help test and improve this guide`

jasonblais replied:
Maybe break this section into two paragraphs, starting with `Once the community has confirmed we have multiple deployments ...`

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+Configure the Firewall
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. note:: This is not required if you are installing on a single server.
```
I'm not sure about the formatting of the note sections, e.g.:

![image](https://cloud.githubusercontent.com/assets/13119842/20434143/8bc318b6-ad74-11e6-91fd-69b005d67357.png)

Wondering if `Note: This is not required if you are installing on a single server.` feels better.

JeffSchering replied:
The ".. note:" is correct for rst docs. The GitHub processor styles it differently than the Sphinx processor that's used for Mattermost docs. This is what it looks like when built using Sphinx:

![image](https://cloud.githubusercontent.com/assets/20325173/20493474/d03d57c6-afcd-11e6-9c87-5f234b647623.png)

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+manually start/stop Mattermost.  The included binary cannot be directly registered as a windows service; therefore, 
+a wrapper-utility must be used.
+
+7. Download the latest version of `NSSM (Non-Sucking Service Manager) <https://nssm.cc/download>`__.
```
I think numbering here should start with "9"

shieldsjared replied:
Indeed, but OBE after removing the requirement to download 7-zip to extract the .tar.gz.

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To verify all steps executed thus far, we'll attempt to access Mattermost through standard HTTP traffic.  To perform 
+this step, you'll need access to a client machine with a compatible browser (e.g. Safari, Firefox, IE11).
```
Is it only these three browsers?

If not, might be better to reference Chrome too, since it's also popular on Windows machines

shieldsjared replied:
Oops!  Changed to include Chrome and "etc"!

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+   
+4. Update **General** > **Configuration** settings to properly configure your reverse proxy by entering `https://mattermost.example.com` as the **Site URL**
+
+   ..alert:: Failure to properly set the Site URL properly __will__ result in unexpected behavior.  Do not include the trailing '/' on the URL.
```
I think the formatting for `..alert::` isn't expected:

![image](https://cloud.githubusercontent.com/assets/13119842/20434297/3973b010-ad75-11e6-84a9-a1911ed0f2d2.png)

jasonblais commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+
+5. Update **Notification** > **Email** settings to setup an SMTP email service. The example below assumes AmazonSES.
+
+   a. Set *Send Email Notifications* to true
```
Bold the field names for settings instead of italicizing

leram84 commented:
If i plan on running all three components on a single machine. Can i substitute all instances of `10.0.0.x` with `127.0.0.1`?

Specifically line 141, can i use:
    "DataSource": "mmuser:mmuser_password@tcp(127.0.0.1:3306)/mattermost?charset=utf8mb4,utf8"

And is there anything else i should know that would be different for my setup:
win10 pro
nginx w ssl
php5.5.9

shieldsjared commented:
You can definitely use localhost throughout. I can't speak to nginx. php should have no impact that I'm aware of. (I'm running it all on a single windows server for production purposes)

leram84 commented:
If someone wants to add an image for lines 104-109 i know i would appreciate it. I'm confused because on 105 `CREATE USER mmuser IDENTIFIED BY 'mmuser_password';` has the password field wrapped in apostrophe's but not the username.... and googling this shows so many different versions that i can't tell if people are using them to try and `insert code blocks` or what... but since i don't really know how to go back and fix this if i screw it up, i would like to get the syntax right the first time. So assuming the database is `mattermost` the user is `mmuser` and the password is `password` can someone just give me the code blocks as if i was going to just copy paste them in.  (lol, sorry, i realize this is a little ridiculous, but at least you get an nginx tester out of it?)  :))

shieldsjared commented:
Love the feedback @leram84! I'll knock it all of the suggestions at once, once ur finished. Thanks again for your help!!!

leram84 commented:
lol, back at you! .... but im actually asking lol. I don't want to trial and error this, so im not going to be finished until you can tell me how to enter that command in :)  Assuming the database is `mattermost` the user is `mmuser` and the password is `password`  Should the command be:
`CREATE USER mmuser IDENTIFIED BY password;`
or
`CREATE USER mmuser IDENTIFIED BY 'password';`
or
`CREATE USER 'mmuser' IDENTIFIED BY 'password';`

shieldsjared commented:
Ah, sorry!

The full answer is "it depends". But for safety and to cover all cases (if somebody decides to use a different password or username with special characters), use the third example with single quotes around the username and password

shieldsjared commented:
But I'm a MSSQL guy, not MySQL so forgive me if I'm wrong!!

leram84 commented:
well actually, you just kinda cleared something else up... i thought that command was actually creating a user called `mmuser` with a password that i set there. i didn't realize it was code to allow you to create usernames elsewhere (at least that's what i think your explanation implies) so i shouldn't actually create a password here, but instead use the `mmuser_password` command just like that. in any case it seems like i can't go wrong with:

`CREATE USER 'mmuser' IDENTIFIED BY 'mmuser_password';`

so that's what ill use.... out for now though, so ill finish this up tomorrow and report back. 

THANKS AGAIN!

leram84 commented:
whats up with the download link? was able to get it on my dev machine w no problem but now going to https://releases.mattermost.com/3.5.0/mattermost-team-3.5.0-windows-amd64.tar.gz doesn't download anything.

shieldsjared commented:
Swap the extension to .zip for the windows download. I'll make sure that gets updated in the guide!

leram84 commented:
Well i can confirm as much of this guide that i used works great! I'm able to run MM locally without issue (so far lol). I'm still stuck with getting this running in my webserver, and have opened an issue [here](https://github.com/mattermost/platform/issues/4607) (if anyone sees this here and is familiar enough with nginx to give me a hand, it would be much appreciated!) But other than the few minor issues i had, this guide is very well done. THANKS FOR ALL YOUR HELP!

JeffSchering commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+   favorite browser and navigating to 
+   ``https://releases.mattermost.com/X.X.X/mattermost-team-X.X.X-windows-amd64.tar.gz`` 
+   where `X.X.X` is the latest Mattermost release version. For example, 3.4.0.
+
```
The link should be ``https://releases.mattermost.com/X.X.X/mattermost-team-X.X.X-windows-amd64.zip`` because the ".tar.gz" does not exist.

If you're going to mention the version, then it needs to be updated every release. And if it's going to be updated every release, then might as well code it directly into the link, so that it becomes ``https://releases.mattermost.com/3.5.0/mattermost-team-3.5.0-windows-amd64.zip``. And if you make it a live link, then people can click and download directly from the docs without having to visit the download page.

JeffSchering commented on source/install/prod-windows-2012.rst:
```diff
@@ -0,0 +1,419 @@
+
+3. Using a utility such as `7-zip <http://www.7-zip.org/>`__, extract the Mattermost Server files 
+   to ``c:\mattermost\``.
+
```
Do we need 7-zip if the download is a zip file?

it33 commented:
@shieldsjared sorry this is taking so long. This is really great work. 

It looks like there's a technical conflict to solve, could you make a tweak and then I'll merge this personally? 

After it's in I'll probably put an "unofficial" label of some kind on it and make sure it's explicitly excluded from our support SLAs until we can build a team around it--and we'll make a decision on how to include it in the table of contents. 

Really fantastic work, let's get this in :)

shieldsjared commented:
Thanks everyone for taking the time to review and provide feedback!!

I've incorporated all feedback, as follows:
- Removed the guide from the TOC (@it33)
- Initial Disclosure: Adjusted grammar broke into two paragraphs. (@jasonblais)
- Updated recommended list of browsers to test with to include Chrome. (@jasonblais)
- Converted italicized field names to bold styling. (@jasonblais)
- Converted download link to use *.zip now that it's officially being built as such (@JeffSchering)
- Removed the need to download a utility to unzip the *.tar.gz (@JeffSchering)
- Converted `alert` admonition to `attention` during the finish-up section
- Removed note the an editor that accomodates unix-style line feeds is necessary (no longer required).
- Added note that when running the all components on a single server, a single IP may be substituted. (@leram84)
- Quoted username identifiers in the mysql statements for clarity (@leram84)
- Various grammer cleanup

@it33 - I did not add an unofficial label, as I wanted to leave that for you to figure out how to best implement that.  I did ensure that the guide will not appear in the TOC.

jasonblais commented:
Fantastic work @shieldsjared!!

it33 commented:
Thanks @shieldsjared! Merged, and I'll add the unofficial piece shortly. 

Outstanding work, thank you so much! This brings the open source project to a new level. 

I was talking this week to Chris Romp from https://r3interworks.com/ who might be able to help test more and work with us on getting this setup officially supported, and then announced, 

Hopefully this will start bringing in contributors from the Windows community as well, which could be pretty huge,

it33 commented:
@shieldsjared @jasonblais I've created and "unofficial enhancements" section under Developer's Guide and added Windows and Video/Audio calling there: https://docs.mattermost.com/guides/developer.html#unofficial-enhancements

I think we should use this as a system in future for these unofficial enhancements, Custom Emoji might go there in future, depending on how far the feature comes by January. It could be a section of our announcements in future as well--and now that we have Windows Server, next is probably SQL Server support
